filer/admin/tools.py

```diff
@@ -54,3 +59,54 @@ def userperms_for_request(item, request):
             if x:
                 r.append(p)
     return r
+
+
+def popup_status(request):
+    return (IS_POPUP_VAR in request.GET or 'pop' in request.GET or
+            IS_POPUP_VAR in request.POST or 'pop' in request.POST)
+
+
+def popup_pick_type(request):
+    # very important to limit the pick_types because the result is marked safe.
+    # (injection attacks)
+    pick_type = request.GET.get('_pick', request.POST.get('_pick'))
+    if pick_type in ALLOWED_PICK_TYPES:
+        return pick_type
+    return None
+
+
+def admin_url_params(request):
+    """
+    given a request, looks at GET and POST values to determine which params
+    should be added. Is used to keep the context of popup and picker mode.
+    """
+    params = {}
+    if popup_status(request):
+        params[IS_POPUP_VAR] = '1'
+    pick_type = popup_pick_type(request)
+    if pick_type:
+        params['_pick'] = pick_type
+    return params
+
+
+def admin_url_params_encoded(request, first_separator='?'):
+    # sorted to make testing easier
+    params = urlencode(sorted(admin_url_params(request).items()))
+    if not params:
+        return ''
+    return '{0}{1}'.format(first_separator, params)
+
+
+class AdminContext(dict):
+    def __init__(self, request):
+        super(AdminContext, self).__init__()
+        self.request = request
+        self.update(admin_url_params(request))
+        extra = dict()
+        extra['popup'] = self.get(IS_POPUP_VAR, False) == '1'
+        extra['pick'] = self.get('_pick', '')
+        for pick_type in ALLOWED_PICK_TYPES:
+            extra['pick_{0}'.format(pick_type)] = extra['pick'] == pick_type
+        for key, value in extra.items():
```

Comment:
wouldn't it be better to override ``__missing__``? setting it as an instance attribute seems a bit hackish, given that if you update the instance using ``instance['key'] = value`` it won't add it as an attribute and this might pose an issue of inconsistent behavior.